### ISSUE TITLE
time_select does not work with :datetime-attributes

### ISSUE THREAD

**Issue by fschwahn** [makandra/active_type#32]
Hi,
I have an issue where `time_select` does not work with `:datetime`-attributes. When passing the params to the initializer (such as `{"start_time(1i)"=>"2015", "start_time(2i)"=>"2", "start_time(3i)"=>"11", "start_time(4i)"=>"16", ....`), rails returns a `ActiveRecord::MultiparameterAssignmentErrors`-exception.

This one just masks another exception, which is triggered here: https://github.com/rails/rails/blob/3-2-stable/activerecord/lib/active_record/attribute_assignment.rb#L149

The expression `(self.class.reflect_on_aggregation(name.to_sym) || column_for_attribute(name))` evaluates to nil, which blows up when `klass` is called on it.

I'm on rails 3.2.21 and active_type 0.3.3. Any idea?

All the best,
Fabian

**Comment by henning-koch** [comment]
Hm, we haven't really any code in place to deal with the very special case of `time_select` helpers (multiple form parameters resulting in a single value), but then again ActiveType classes inherit from ActiveRecord and should inherit AR's behavior.

**Comment by fschwahn** [comment]
Yes, I was sure that this would just work without any special code because of your approach. I looked quickly at the rails source, and `column_for_attribute` just looks at the class attribute `columns_hash` and returns the klass defined for it. So adding the following to my class made it work:

```ruby
self.columns_hash['start_time'] = OpenStruct.new(klass: Time)
```

I'm not deep in ActiveRecord, so I don't know what other implications that may have, but if there are no side effects it may make sense to populate the `columns_hash` when attributes are defined.

**Comment by henning-koch** [comment]
Can you confirm this is **not** an issue with ActiveRecord? ActiveType only minics ActiveRecord behavior.

**Comment by fschwahn** [comment]
Yes, I'm sure it is an issue with ActiveType. You add your own `virtual_columns_hash` while `time_select` uses `column_for_attribute`, which expect `columns_hash` to be populated. This issue can probably be fixed by overriding the `column_for_attribute`-method (https://github.com/rails/rails/blob/3-2-stable/activerecord/lib/active_record/attribute_methods.rb#L222) to also check `virtual_columns_hash` for the passed column. `VirtualColumn` would need to implement a `klass`-method for it to work, I'm not sure if it does that right now.

For me, the workaround I posted above works for now.

**Comment by renato0307** [comment]
Hi,

I'm having a similar issue with Rails 4.2.1:

```
1 error(s) on assignment of multiparameter attributes [error on assignment [1, 2, 2011] to new_payed_at (undefined method `klass' for nil:NilClass)]
```

I did some investigation and found the problem to be occurring in the following line:
https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/attribute_assignment.rb#L130

After reading part of the activerecord code I was able to overcome the problem by extending some of the active_type classes:
```ruby
class ActiveType::Object
  def type_for_attribute(attribute)    
    virtual_columns_hash[attribute]
  end
end

class ActiveType::VirtualAttributes::VirtualColumn
  def type
    @type_caster.type
  end

  def klass
    case type
      when :integer                     then Fixnum
      when :float                       then Float
      when :decimal                     then BigDecimal
      when :datetime, :timestamp, :time then Time
      when :date                        then Date
      when :text, :string, :binary      then String
      when :boolean                     then Object
    end
  end  
end

class ActiveType::TypeCaster
  attr_reader :type
end
```

As I'm new to rails, ruby and active_type this might not be the best solution.

Any ideia or comments?

Thank you in advance.

Renato
